In [1]:
import sys
import os

sys.path.append(os.path.abspath(".."))

In [2]:
import os
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader

from models.attention_unet_3d import AttentionUNet3D
from utils.unets_helper_functions import (
    set_seed,
    train_one_epoch,
    validate_one_epoch,
    save_checkpoint,
    PatchDataset,
    evaluate_full_volume
)


In [3]:
set_seed(42)

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print("Using device:", device)

Using device: cuda


In [4]:
FOLD = 0

with open(f"../data/splits/fold_{FOLD}/train.txt") as f:
    train_cases = f.read().splitlines()

with open(f"../data/splits/fold_{FOLD}/val.txt") as f:
    val_cases = f.read().splitlines()

print("Train cases:", len(train_cases))
print("Val cases:", len(val_cases))

Train cases: 33
Val cases: 9


In [5]:
PATCH_SIZE = 80
PATCHES_PER_CASE = 6

train_dataset = PatchDataset(train_cases,
                        "../data/processed/imagesTr",
                        "../data/processed/labelsTr",
                        patches_per_case = PATCHES_PER_CASE,
                        patch_size = PATCH_SIZE,
                        augment=False)

val_dataset = PatchDataset(val_cases,
                        "../data/processed/imagesTr",
                        "../data/processed/labelsTr",
                        patches_per_case = 1,
                        patch_size = PATCH_SIZE,
                        augment=False)


train_loader = DataLoader(
    train_dataset,
    batch_size=2,
    shuffle=True,
    num_workers=2,      
    pin_memory=True    
)

val_loader = DataLoader(
    val_dataset,
    batch_size=1,
    num_workers=2,    
    pin_memory=True
)

In [6]:
model = AttentionUNet3D(
    in_channels=1,
    out_channels=7,
    base_filters=16
).to(device)

print("Model initialized.")

Model initialized.


In [7]:
optimizer = optim.Adam(model.parameters(), lr=1e-4)
weights = torch.tensor([0.1, 1, 1, 1, 1, 1, 1]).to(device)
ce_loss = nn.CrossEntropyLoss(weight=weights)
scaler = torch.cuda.amp.GradScaler()

C:\Users\dhanu\AppData\Local\Temp\ipykernel_8936\1548250201.py:4: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  scaler = torch.cuda.amp.GradScaler()


In [8]:
EPOCHS = 40

PROJECT_ROOT = os.path.abspath("..")
SAVE_DIR = os.path.join(PROJECT_ROOT, "experiments", "attention_unet_fold0")

os.makedirs(SAVE_DIR, exist_ok=True)
best_val_loss = float("inf")

for epoch in range(EPOCHS):

    train_loss = train_one_epoch(
        model,
        train_loader,
        optimizer,
        scaler,
        ce_loss,
        device
    )

    val_loss, val_dice = validate_one_epoch(
        model,
        val_loader,
        ce_loss,
        device
    )

    print(f"\nEpoch {epoch}")
    print(f"Train Loss: {train_loss:.4f}")
    print(f"Val Loss: {val_loss:.4f}")
    print(f"Per Class Dice: {val_dice}")

    # Save best model
    if val_loss < best_val_loss:
        best_val_loss = val_loss
        save_checkpoint(
            model,
            optimizer,
            epoch,
            best_val_loss,
            os.path.join(SAVE_DIR, "best_model.pth")
        )
        print("Saved Best Model")

    # Save checkpoint every 10 epochs
    if (epoch + 1) % 10 == 0:
        save_checkpoint(
            model,
            optimizer,
            epoch,
            best_val_loss,
            os.path.join(SAVE_DIR, f"checkpoint_epoch_{epoch+1}.pth")
        )
        print("Checkpoint Saved")

100%|██████████| 99/99 [04:34<00:00,  2.77s/it]



Epoch 0
Train Loss: 2.8346
Val Loss: 2.6008
Per Class Dice: [2.42432162e-01 5.91081654e-03 4.91796294e-08 7.07735789e-03
 3.73778992e-03 1.83764201e-01]
Saved Best Model


100%|██████████| 99/99 [04:10<00:00,  2.53s/it]



Epoch 1
Train Loss: 2.5491
Val Loss: 2.4636
Per Class Dice: [2.90280402e-01 1.86149233e-02 4.59095763e-07 1.17944456e-02
 7.17534182e-03 2.78749436e-01]
Saved Best Model


100%|██████████| 99/99 [03:41<00:00,  2.24s/it]



Epoch 2
Train Loss: 2.4825
Val Loss: 2.5055
Per Class Dice: [0.22095573 0.03520555 0.00046623 0.02909724 0.00616944 0.13061954]


100%|██████████| 99/99 [04:00<00:00,  2.43s/it]



Epoch 3
Train Loss: 2.3215
Val Loss: 2.3197
Per Class Dice: [3.02377234e-01 4.86563486e-02 9.59373282e-08 3.81717894e-03
 6.89568363e-03 2.41302282e-01]
Saved Best Model


100%|██████████| 99/99 [04:04<00:00,  2.47s/it]



Epoch 4
Train Loss: 2.2228
Val Loss: 2.1385
Per Class Dice: [0.32311175 0.14002334 0.33333341 0.09130889 0.01064318 0.34125892]
Saved Best Model


100%|██████████| 99/99 [04:19<00:00,  2.62s/it]



Epoch 5
Train Loss: 2.1518
Val Loss: 2.1628
Per Class Dice: [0.28486497 0.19177718 0.33333349 0.0575127  0.01985317 0.26039359]


100%|██████████| 99/99 [04:06<00:00,  2.49s/it]



Epoch 6
Train Loss: 2.0526
Val Loss: 2.0826
Per Class Dice: [0.27703481 0.2045845  0.33333355 0.06555741 0.00933331 0.2505439 ]
Saved Best Model


100%|██████████| 99/99 [04:02<00:00,  2.45s/it]



Epoch 7
Train Loss: 2.0000
Val Loss: 1.9149
Per Class Dice: [0.43775315 0.2635214  0.44444572 0.1070518  0.0036978  0.40621116]
Saved Best Model


100%|██████████| 99/99 [04:09<00:00,  2.52s/it]



Epoch 8
Train Loss: 1.8924
Val Loss: 1.9311
Per Class Dice: [0.24737168 0.25563065 0.77777778 0.18871078 0.33773419 0.2445866 ]


100%|██████████| 99/99 [03:59<00:00,  2.42s/it]



Epoch 9
Train Loss: 1.8349
Val Loss: 1.8353
Per Class Dice: [0.39735735 0.330353   0.66666667 0.19472881 0.27574055 0.30859587]
Saved Best Model
Checkpoint Saved


100%|██████████| 99/99 [04:00<00:00,  2.43s/it]



Epoch 10
Train Loss: 1.7794
Val Loss: 1.7888
Per Class Dice: [0.37398141 0.26481382 0.44444447 0.25943298 0.4688684  0.40940163]
Saved Best Model


100%|██████████| 99/99 [03:47<00:00,  2.30s/it]



Epoch 11
Train Loss: 1.7303
Val Loss: 1.6762
Per Class Dice: [0.53426427 0.44054092 0.55555556 0.32535697 0.37702557 0.57650394]
Saved Best Model


100%|██████████| 99/99 [04:00<00:00,  2.43s/it]



Epoch 12
Train Loss: 1.6278
Val Loss: 1.6499
Per Class Dice: [0.46286663 0.62848407 0.44444445 0.36023861 0.26169391 0.41979846]
Saved Best Model


100%|██████████| 99/99 [03:56<00:00,  2.39s/it]



Epoch 13
Train Loss: 1.6011
Val Loss: 1.5485
Per Class Dice: [0.50607414 0.69734662 0.44444445 0.22814109 0.32371176 0.52907269]
Saved Best Model


100%|██████████| 99/99 [04:05<00:00,  2.48s/it]



Epoch 14
Train Loss: 1.5432
Val Loss: 1.5285
Per Class Dice: [0.48676583 0.45515373 0.66666667 0.21899413 0.46105628 0.33760718]
Saved Best Model


100%|██████████| 99/99 [03:55<00:00,  2.38s/it]



Epoch 15
Train Loss: 1.5224
Val Loss: 1.6058
Per Class Dice: [0.19166938 0.80616955 0.77777778 0.48802255 0.58623931 0.10739835]


100%|██████████| 99/99 [04:09<00:00,  2.52s/it]



Epoch 16
Train Loss: 1.4738
Val Loss: 1.4437
Per Class Dice: [0.43816758 0.88444941 0.55555557 0.5568583  0.5605536  0.4001902 ]
Saved Best Model


100%|██████████| 99/99 [04:04<00:00,  2.47s/it]



Epoch 17
Train Loss: 1.4273
Val Loss: 1.3935
Per Class Dice: [0.47863889 0.92504667 0.5555557  0.2991313  0.60377436 0.57335543]
Saved Best Model


100%|██████████| 99/99 [04:07<00:00,  2.50s/it]



Epoch 18
Train Loss: 1.3697
Val Loss: 1.3424
Per Class Dice: [0.5555288  0.72757239 0.55555558 0.66224666 0.74735402 0.54981383]
Saved Best Model


100%|██████████| 99/99 [03:57<00:00,  2.40s/it]



Epoch 19
Train Loss: 1.3449
Val Loss: 1.3166
Per Class Dice: [0.73255518 0.6891863  0.22222224 0.71538286 0.81508009 0.64050299]
Saved Best Model
Checkpoint Saved


100%|██████████| 99/99 [03:53<00:00,  2.36s/it]



Epoch 20
Train Loss: 1.3185
Val Loss: 1.3420
Per Class Dice: [0.41088164 0.75257575 0.77777778 0.37235488 0.55555612 0.39351219]


100%|██████████| 99/99 [04:00<00:00,  2.43s/it]



Epoch 21
Train Loss: 1.2691
Val Loss: 1.2270
Per Class Dice: [0.79729281 0.81408644 0.55555556 0.75445729 0.8605074  0.78714084]
Saved Best Model


100%|██████████| 99/99 [04:06<00:00,  2.49s/it]



Epoch 22
Train Loss: 1.2399
Val Loss: 1.2314
Per Class Dice: [0.58130454 0.77000625 0.33333334 0.54609746 0.46472496 0.50368574]


100%|██████████| 99/99 [03:52<00:00,  2.34s/it]



Epoch 23
Train Loss: 1.2397
Val Loss: 1.1680
Per Class Dice: [0.65403311 0.8191313  0.66666668 0.4404329  0.6673     0.5928562 ]
Saved Best Model


100%|██████████| 99/99 [04:05<00:00,  2.48s/it]



Epoch 24
Train Loss: 1.2048
Val Loss: 1.2112
Per Class Dice: [0.69353057 0.87578168 0.88888889 0.7194252  0.77321634 0.61801043]


100%|██████████| 99/99 [04:07<00:00,  2.50s/it]



Epoch 25
Train Loss: 1.1785
Val Loss: 1.1465
Per Class Dice: [0.74344082 0.81147252 0.33333334 0.83363685 0.72192039 0.76673326]
Saved Best Model


100%|██████████| 99/99 [03:56<00:00,  2.39s/it]



Epoch 26
Train Loss: 1.1660
Val Loss: 1.1304
Per Class Dice: [0.72233706 0.82933034 0.33333334 0.45831228 0.55159142 0.67567516]
Saved Best Model


100%|██████████| 99/99 [03:54<00:00,  2.37s/it]



Epoch 27
Train Loss: 1.1456
Val Loss: 1.1488
Per Class Dice: [0.75714915 0.91584918 0.44444445 0.5995495  0.72734356 0.780385  ]


100%|██████████| 99/99 [04:14<00:00,  2.57s/it]



Epoch 28
Train Loss: 1.0891
Val Loss: 1.0975
Per Class Dice: [0.78294841 0.77920853 0.66666667 0.61305339 0.47952179 0.74433115]
Saved Best Model


100%|██████████| 99/99 [03:45<00:00,  2.28s/it]



Epoch 29
Train Loss: 1.1354
Val Loss: 1.0529
Per Class Dice: [0.74073419 0.71470977 0.33333334 0.46762842 0.83607649 0.66291256]
Saved Best Model
Checkpoint Saved


100%|██████████| 99/99 [03:48<00:00,  2.31s/it]



Epoch 30
Train Loss: 1.1029
Val Loss: 1.0782
Per Class Dice: [0.75117918 0.64604946 0.5555556  0.47135369 0.62598981 0.60486718]


100%|██████████| 99/99 [04:12<00:00,  2.55s/it]



Epoch 31
Train Loss: 1.0740
Val Loss: 1.1184
Per Class Dice: [0.54326764 0.53728164 0.66666667 0.60421671 0.54045328 0.61217794]


100%|██████████| 99/99 [04:03<00:00,  2.46s/it]



Epoch 32
Train Loss: 1.0534
Val Loss: 0.9942
Per Class Dice: [0.8987946  0.74179398 0.33333362 0.75871551 0.54943452 0.8395661 ]
Saved Best Model


100%|██████████| 99/99 [04:16<00:00,  2.59s/it]



Epoch 33
Train Loss: 1.0358
Val Loss: 1.0064
Per Class Dice: [0.86650942 0.8276382  0.44444445 0.73511541 0.76485298 0.70048094]


100%|██████████| 99/99 [04:15<00:00,  2.58s/it]



Epoch 34
Train Loss: 1.0320
Val Loss: 1.0470
Per Class Dice: [0.59431522 0.81125796 0.44444445 0.61822207 0.687052   0.65429486]


100%|██████████| 99/99 [03:59<00:00,  2.42s/it]



Epoch 35
Train Loss: 1.0225
Val Loss: 1.0241
Per Class Dice: [0.7115901  0.89329365 0.33333334 0.62108979 0.76920573 0.64803393]


100%|██████████| 99/99 [04:02<00:00,  2.45s/it]



Epoch 36
Train Loss: 1.0184
Val Loss: 1.0771
Per Class Dice: [0.55068955 0.68595257 0.88888889 0.64335502 0.84210847 0.67217312]


100%|██████████| 99/99 [04:11<00:00,  2.54s/it]



Epoch 37
Train Loss: 0.9957
Val Loss: 1.0254
Per Class Dice: [0.61559095 0.57658315 0.55555567 0.77156627 0.43418022 0.58131144]


100%|██████████| 99/99 [04:02<00:00,  2.45s/it]



Epoch 38
Train Loss: 0.9845
Val Loss: 1.0361
Per Class Dice: [0.6137176  0.80674572 0.66666667 0.64306937 0.71762863 0.56444355]


100%|██████████| 99/99 [04:04<00:00,  2.47s/it]



Epoch 39
Train Loss: 0.9775
Val Loss: 0.9916
Per Class Dice: [0.82103864 0.89958535 0.55555556 0.63425321 0.61709103 0.52578725]
Saved Best Model
Checkpoint Saved


In [5]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

model = AttentionUNet3D().to(device)

model_path = "../experiments/attention_unet_fold0/best_model.pth"
checkpoint = torch.load(model_path, map_location=device)
model.load_state_dict(checkpoint["model_state_dict"])

model.eval()

print("Model loaded for testing.")


C:\Users\dhanu\AppData\Local\Temp\ipykernel_28492\3142351894.py:6: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  checkpoint = torch.load(model_path, map_location=device)


Model loaded for testing.


In [11]:
print("Training Complete")

Training Complete


In [6]:
mean_dice, std_dice = evaluate_full_volume(
    model,
    val_cases,
    "../data/processed/imagesTr",
    "../data/processed/labelsTr",
    device=device
)

Unique pred labels: [0 1 2 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_38 Dice: [np.float64(0.8818314276833449), np.float64(0.7960853794009235), np.float64(3.752345201679755e-09), np.float64(0.709722115534045), np.float64(0.7790560741198876), np.float64(0.8711903700630957)]
Unique pred labels: [0 1 2 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_05 Dice: [np.float64(0.8232048513069763), np.float64(0.6526627619654866), np.float64(3.4340659222731253e-09), np.float64(0.6981543252398691), np.float64(0.7576708861986476), np.float64(0.8358394848209485)]
Unique pred labels: [0 1 2 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_22 Dice: [np.float64(0.8291557480263313), np.float64(0.7853061225837952), np.float64(3.409478338187936e-09), np.float64(0.6997645829454541), np.float64(0.8087727612477), np.float64(0.878043593851592)]
Unique pred labels: [0 1 2 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_14 Dice: [np.float64(0.8099049525290686), np.float64(0.7444335204762977), np.float64(4.4964028574

## Attention UNet 3D Patch-Based Validation Performance

The model was trained for 40 epochs using patch-based validation.  
Below are the final patch validation metrics obtained at the best validation loss epoch.

### Final Training Statistics (Epoch 39 - Best Model)

- Train Loss: 0.9775  
- Validation Loss: 0.9916  

### Patch-Based Per-Class Dice (Validation)

- Mandible: 0.8210  
- Brainstem: 0.8996  
- Spinal Cord: 0.5556  
- Parotid Left: 0.6343  
- Parotid Right: 0.6171  
- Oral Cavity: 0.5258  


## Observations

- Brainstem achieved very strong performance (~0.90 Dice).
- Mandible segmentation is robust (>0.82 Dice).
- Parotid glands show moderate performance (~0.61–0.63).
- Spinal Cord performance is moderate (~0.56).
- Oral Cavity segmentation remains challenging.

Overall, the Attention UNet demonstrates strong patch-based segmentation performance, particularly for Mandible and Brainstem.


# Evaluating on Full Volume

~ (patch = 80 , stride = 60)

| Organ ID | Organ (Based on Your Map) | Mean Dice   | Std Dev |
| -------- | ------------------------- | ----------- | ------- |
| 1        | Bone Mandible             | **0.8425**  | 0.0365  |
| 2        | Brainstem                 | **0.7655**  | 0.0532  |
| 3        | Spinal Cord               | **~0.0000** | ~0.0000 |
| 4        | Parotid Left              | **0.6701**  | 0.0585  |
| 5        | Parotid Right             | **0.7425**  | 0.0841  |
| 6        | Oral Cavity               | **0.7889**  | 0.0839  |


Full Volume Evaluation Summary

**Strong Organs**

- Mandible (0.84) → Excellent and very stable
- Brainstem (0.77) → Strong and consistent
- Oral Cavity (0.79) → Significantly improved compared to UNet++
- Parotid Right (0.74) → Good bilateral performance

**Moderate**
- Parotid Left (0.67)
- Slightly lower but still consistent across cases
- Failure Case

- Spinal Cord (~0.00)
The model failed to predict class 3 in full-volume inference.
This suggests:
Extreme class imbalance sensitivity
Thin elongated structure suppression during sliding-window blending
Attention mechanism possibly suppressing weak spinal features